<a href="https://colab.research.google.com/github/rubensolano2/GUIDES-AND-PRACTICES/blob/main/GUIDE%20OF%20MACHINE%20LEARNING.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#LIBRERIAS

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

- numpy: Es una biblioteca muy popular en Python que se utiliza para realizar cálculos numéricos y científicos en general. Proporciona estructuras de datos para trabajar con matrices y vectores, así como funciones matemáticas para realizar operaciones en estos objetos.
En este caso, numpy se importa con el alias np, lo que significa que podemos acceder a las funciones y estructuras de datos de numpy utilizando np en lugar de numpy. Por ejemplo, si queremos crear un arreglo de números, podemos hacerlo utilizando la función np.array() en lugar de numpy.array().

- pandas: Es otra biblioteca popular de Python que se utiliza para manipular y analizar datos en formato tabular (como una hoja de cálculo). pandas proporciona estructuras de datos de alto nivel como el DataFrame, que permite organizar datos en filas y columnas, y funciones para realizar operaciones en estos objetos, como la agregación y el filtrado.
En este caso, pandas se importa con el alias pd, lo que significa que podemos acceder a las funciones y estructuras de datos de pandas utilizando pd en lugar de pandas. Por ejemplo, si queremos leer un archivo de datos CSV y cargarlo en un DataFrame, podemos hacerlo utilizando la función pd.read_csv() en lugar de pandas.read_csv().

- matplotlib: Es una biblioteca de visualización de datos en Python que se utiliza para crear gráficos y visualizaciones. matplotlib proporciona una amplia variedad de herramientas para crear gráficos de diferentes tipos, como gráficos de línea, barras, histogramas, diagramas de dispersión, etc.
En este caso, matplotlib se importa con el alias plt, lo que significa que podemos acceder a las funciones y estructuras de datos de matplotlib utilizando plt en lugar de matplotlib. Por ejemplo, si queremos crear un gráfico de línea simple, podemos hacerlo utilizando la función plt.plot() en lugar de matplotlib.pyplot.plot().

#importacion dataset y limpieza de datos

- La importación de un dataset se refiere al proceso de cargar datos de un archivo o una fuente externa en un programa de análisis de datos. En Python, esto se puede hacer utilizando bibliotecas como pandas, que proporciona funciones para leer y cargar datos en un DataFrame.

- El proceso de limpieza de datos se refiere a la tarea de identificar y corregir errores, inconsistencias y datos faltantes en el dataset para que pueda ser utilizado de manera efectiva en el análisis. La limpieza de datos es una parte esencial del proceso de análisis de datos, ya que los datos limpios y precisos son fundamentales para tomar decisiones precisas e informadas.

In [ ]:
# Importar el dataset Iris desde scikit-learn
from sklearn import datasets

# Cargar el dataset Iris
iris = datasets.load_iris()
X = iris.data
y = iris.target

##info del dataset:

-  El conjunto de datos Iris es un conjunto de datos clásico en el campo de Machine Learning y estadísticas. Consiste en 150 instancias de flores Iris, distribuidas equitativamente entre tres especies: Iris setosa, Iris versicolor e Iris virginica (50 instancias de cada especie). Cada instancia contiene cuatro características, que son las medidas del sépalo y pétalo de cada flor (longitud y anchura del sépalo, longitud y anchura del pétalo). El objetivo es clasificar las flores en una de las tres especies a partir de estas características.

- Cuando se carga el conjunto de datos Iris desde la biblioteca scikit-learn, se obtiene un objeto tipo 'Bunch'. Este objeto es similar a un diccionario y contiene varios atributos útiles para trabajar con el conjunto de datos:

- **iris.data**: Es un array bidimensional de tamaño (150, 4), donde cada fila representa una instancia de flor Iris y cada columna representa una característica (longitud del sépalo, anchura del sépalo, longitud del pétalo y anchura del pétalo). Es decir, contiene las variables independientes (predictores) del conjunto de datos.

- **iris.target**: Es un array unidimensional de tamaño (150,), donde cada elemento representa la especie de la flor Iris correspondiente a cada instancia en iris.data. Los valores son enteros en el rango 0-2, y cada número representa una especie (0: Iris setosa, 1: Iris versicolor, 2: Iris virginica). Es decir, contiene la variable dependiente (objetivo o etiqueta) del conjunto de datos.
- El conjunto de datos Iris es ampliamente utilizado en el aprendizaje de Machine Learning, especialmente en tareas de clasificación supervisada, porque es simple, fácil de entender y permite a los principiantes familiarizarse con los conceptos básicos de Machine Learning sin tener que preocuparse por la complejidad del conjunto de datos. Además, sus características numéricas y baja dimensionalidad facilitan la visualización y el análisis exploratorio de datos.

In [ ]:
iris

{'data': array([[5.1, 3.5, 1.4, 0.2],
        [4.9, 3. , 1.4, 0.2],
        [4.7, 3.2, 1.3, 0.2],
        [4.6, 3.1, 1.5, 0.2],
        [5. , 3.6, 1.4, 0.2],
        [5.4, nan, 1.7, 0.4],
        [4.6, nan, 1.4, 0.3],
        [5. , nan, 1.5, 0.2],
        [4.4, nan, 1.4, 0.2],
        [4.9, nan, 1.5, 0.1],
        [5.4, 3.7, 1.5, 0.2],
        [4.8, 3.4, 1.6, 0.2],
        [4.8, 3. , 1.4, 0.1],
        [4.3, 3. , 1.1, 0.1],
        [5.8, 4. , 1.2, 0.2],
        [5.7, 4.4, 1.5, 0.4],
        [5.4, 3.9, 1.3, 0.4],
        [5.1, 3.5, 1.4, 0.3],
        [5.7, 3.8, 1.7, 0.3],
        [5.1, 3.8, 1.5, 0.3],
        [5.4, 3.4, 1.7, nan],
        [5.1, 3.7, 1.5, nan],
        [4.6, 3.6, 1. , nan],
        [5.1, 3.3, 1.7, nan],
        [4.8, 3.4, 1.9, nan],
        [5. , 3. , 1.6, 0.2],
        [5. , 3.4, 1.6, 0.4],
        [5.2, 3.5, 1.5, 0.2],
        [5.2, 3.4, 1.4, 0.2],
        [4.7, 3.2, 1.6, 0.2],
        [4.8, 3.1, 1.6, 0.2],
        [5.4, 3.4, 1.5, 0.4],
        [5.2, 4.1, 1.5, 0.1],
  

##EJEMPLO DE LIMPIEZA DE DATOS GENÉRICO

In [ ]:
# Convertir a un DataFrame de pandas
iris_df = pd.DataFrame(X, columns=iris.feature_names)
iris_df['species'] = y

# Asumir que algunas instancias tienen datos faltantes
iris_df.iloc[5:10, 1] = np.nan
iris_df.iloc[20:25, 3] = np.nan

# Verificar si hay datos faltantes
print(iris_df.isnull().sum())

# 1. Eliminar instancias con datos faltantes
iris_df_no_missing = iris_df.dropna()
print("\nDespués de eliminar instancias con datos faltantes:")
print(iris_df_no_missing.isnull().sum())

# 2. Rellenar datos faltantes con la media de la columna
iris_df_filled = iris_df.fillna(iris_df.mean())
print("\nDespués de rellenar datos faltantes con la media de la columna:")
print(iris_df_filled.isnull().sum())

# 3. Eliminar duplicados
iris_df_no_duplicates = iris_df_filled.drop_duplicates()
print("\nDespués de eliminar duplicados, el tamaño del DataFrame es:", iris_df_no_duplicates.shape)


sepal length (cm)    0
sepal width (cm)     5
petal length (cm)    0
petal width (cm)     5
species              0
dtype: int64

Después de eliminar instancias con datos faltantes:
sepal length (cm)    0
sepal width (cm)     0
petal length (cm)    0
petal width (cm)     0
species              0
dtype: int64

Después de rellenar datos faltantes con la media de la columna:
sepal length (cm)    0
sepal width (cm)     0
petal length (cm)    0
petal width (cm)     0
species              0
dtype: int64

Después de eliminar duplicados, el tamaño del DataFrame es: (149, 5)


##TRUCO

Algunas veces te encontrarás con tener que hacer muchos cambios a un dataset un truco para no estar ejecutando todo el código sin para es hacerle una copia al dataset en un momento específico

In [ ]:
iris_copia =iris.copy()

In [ ]:
iris_copia

{'data': array([[5.1, 3.5, 1.4, 0.2],
        [4.9, 3. , 1.4, 0.2],
        [4.7, 3.2, 1.3, 0.2],
        [4.6, 3.1, 1.5, 0.2],
        [5. , 3.6, 1.4, 0.2],
        [5.4, nan, 1.7, 0.4],
        [4.6, nan, 1.4, 0.3],
        [5. , nan, 1.5, 0.2],
        [4.4, nan, 1.4, 0.2],
        [4.9, nan, 1.5, 0.1],
        [5.4, 3.7, 1.5, 0.2],
        [4.8, 3.4, 1.6, 0.2],
        [4.8, 3. , 1.4, 0.1],
        [4.3, 3. , 1.1, 0.1],
        [5.8, 4. , 1.2, 0.2],
        [5.7, 4.4, 1.5, 0.4],
        [5.4, 3.9, 1.3, 0.4],
        [5.1, 3.5, 1.4, 0.3],
        [5.7, 3.8, 1.7, 0.3],
        [5.1, 3.8, 1.5, 0.3],
        [5.4, 3.4, 1.7, nan],
        [5.1, 3.7, 1.5, nan],
        [4.6, 3.6, 1. , nan],
        [5.1, 3.3, 1.7, nan],
        [4.8, 3.4, 1.9, nan],
        [5. , 3. , 1.6, 0.2],
        [5. , 3.4, 1.6, 0.4],
        [5.2, 3.5, 1.5, 0.2],
        [5.2, 3.4, 1.4, 0.2],
        [4.7, 3.2, 1.6, 0.2],
        [4.8, 3.1, 1.6, 0.2],
        [5.4, 3.4, 1.5, 0.4],
        [5.2, 4.1, 1.5, 0.1],
  

#ESCALADO DE DATOS

El escalado de datos es una técnica de preprocesamiento que se utiliza para poner todas las características en la misma escala, normalmente en el rango [0, 1] o con media 0 y varianza 1. Esto es importante porque algunos algoritmos de Machine Learning son sensibles a la escala de las características y pueden tener un rendimiento deficiente si las características están en escalas muy diferentes.

A continuación, se muestra un ejemplo de cómo escalar los datos del conjunto de datos Iris utilizando la biblioteca scikit-learn:



INFO ADICIONAL:

- Media 0: SIGNIFICA QUE SI HACES LA MEDIA ES DECIR SUMAS TODO Y LO DIVIDES ENTRE EL NUMERO DE DATOS DA 0.

- Varianza 1: La varianza es una medida estadística que indica cuánto se extienden los valores de un conjunto de datos alrededor de su media. Una varianza de 1 indica que los valores de los datos están bastante dispersos en relación a su media.

De manera más específica, la varianza se calcula como la suma de las diferencias al cuadrado entre cada valor y la media, dividido por el número de valores en el conjunto de datos.

In [ ]:
from sklearn import datasets
from sklearn.preprocessing import MinMaxScaler, StandardScaler

# Cargar el dataset Iris
iris = datasets.load_iris()
X = iris.data

# Escalado Min-Max
min_max_scaler = MinMaxScaler()
X_min_max_scaled = min_max_scaler.fit_transform(X)

print("Datos escalados utilizando Min-Max Scaler (primeras 5 instancias):\n", X_min_max_scaled[:5])

# Escalado de características con media 0 y varianza 1 (Standard Scaler)
standard_scaler = StandardScaler()
X_standard_scaled = standard_scaler.fit_transform(X)

print("\nDatos escalados utilizando Standard Scaler (primeras 5 instancias):\n", X_standard_scaled[:5])


Datos escalados utilizando Min-Max Scaler (primeras 5 instancias):
 [[0.22222222 0.625      0.06779661 0.04166667]
 [0.16666667 0.41666667 0.06779661 0.04166667]
 [0.11111111 0.5        0.05084746 0.04166667]
 [0.08333333 0.45833333 0.08474576 0.04166667]
 [0.19444444 0.66666667 0.06779661 0.04166667]]

Datos escalados utilizando Standard Scaler (primeras 5 instancias):
 [[-0.90068117  1.01900435 -1.34022653 -1.3154443 ]
 [-1.14301691 -0.13197948 -1.34022653 -1.3154443 ]
 [-1.38535265  0.32841405 -1.39706395 -1.3154443 ]
 [-1.50652052  0.09821729 -1.2833891  -1.3154443 ]
 [-1.02184904  1.24920112 -1.34022653 -1.3154443 ]]


Explicación:

1. Primero, cargamos el conjunto de datos Iris y extraemos las características en la variable X.

2. Luego, aplicamos el escalado Min-Max utilizando la clase MinMaxScaler de scikit-learn. Este escalador transforma las características escalándolas al rango [0, 1]. Lo hace restando el valor mínimo y dividiendo por el rango (máximo - mínimo) para cada característica. La fórmula es la siguiente:

In [ ]:
X_min_max_scaled = (X - X_min) / (X_max - X_min)

Utilizamos fit_transform para calcular los valores mínimos y máximos de cada característica y aplicar la transformación en un solo paso. Luego, mostramos las primeras 5 instancias de los datos escalados.

3. Después, aplicamos el escalado de características con media 0 y varianza 1 utilizando la clase StandardScaler de scikit-learn. Este escalador transforma las características restando la media y dividiendo por la desviación estándar de cada característica. La fórmula es la siguiente:

In [ ]:
X_standard_scaled = (X - X_mean) / X_std

Utilizamos fit_transform para calcular la media y la desviación estándar de cada característica y aplicar la transformación en un solo paso. Luego, mostramos las primeras 5 instancias de los datos escalados.

Al escalar los datos, asegúrate de ajustar el escalador solo en el conjunto de entrenamiento y aplicarlo tanto al conjunto de entrenamiento como al de prueba. Esto evita la filtración de información del conjunto de prueba durante el proceso de preprocesamiento.

#DIVISIÓN PARA EL ENTRENAMIENTO

- Dividir los datos en conjuntos de entrenamiento y prueba es un paso esencial en el proceso de Machine Learning. La idea principal es separar una parte de los datos para entrenar el modelo y otra parte para evaluar su rendimiento. Esto nos permite estimar cómo de bien el modelo generalizará a datos nunca vistos y nos ayuda a detectar posibles problemas como el sobreajuste (overfitting).

- El sobreajuste ocurre cuando un modelo se ajusta demasiado bien a los datos de entrenamiento y tiene un rendimiento pobre en datos no vistos. Al tener un conjunto de prueba separado, podemos obtener una estimación imparcial del rendimiento del modelo en datos no vistos.

A continuación, te explico cómo dividir los datos en conjuntos de entrenamiento y prueba utilizando la biblioteca scikit-learn:

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn import datasets

# Cargar el dataset Iris
iris = datasets.load_iris()
X = iris.data
y = iris.target

# Dividir los datos en conjuntos de entrenamiento (80%) y prueba (20%)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Mostrar el tamaño de los conjuntos de entrenamiento y prueba
print("Tamaño del conjunto de entrenamiento:", X_train.shape, y_train.shape)
print("Tamaño del conjunto de prueba:", X_test.shape, y_test.shape)

Tamaño del conjunto de entrenamiento: (120, 4) (120,)
Tamaño del conjunto de prueba: (30, 4) (30,)


Explicación:

1. Importamos la función **train_test_split** de scikit-learn, que nos permite dividir fácilmente los datos en conjuntos de entrenamiento y prueba.

2. Cargamos el conjunto de datos Iris y extraemos las características en la variable X y las etiquetas en la variable y.

3. Utilizamos la función **train_test_split** para dividir los datos en conjuntos de entrenamiento y prueba. Establecemos el parámetro test_size en 0.2, lo que significa que el 20% de los datos se reservará para el conjunto de prueba y el 80% restante se utilizará para entrenar el modelo. El parámetro random_state es un número entero que se utiliza como semilla para la generación de números aleatorios. Establecer un valor fijo garantiza que la división sea reproducible en diferentes ejecuciones.

4. La función **train_test_split** devuelve cuatro arrays: **X_train, X_test, y_train y y_test**. Estos representan las características y etiquetas de los conjuntos de entrenamiento y prueba, respectivamente. Mostramos sus dimensiones para verificar la división.

Una vez que hayas dividido los datos, puedes proceder a realizar el preprocesamiento, entrenar el modelo utilizando el conjunto de entrenamiento y evaluar su rendimiento utilizando el conjunto de prueba.

info adicional:

- son vectores realmente, el primero de 120 datos y 4 tipos y el segundo unidimensioanl de 120 datos.

Lo que hacemos es probar sin parar hasta dar con una ecuación que se ajuste de forma correcta a los datos para hacer las predicciones

#VAMOS A MONTAR LOS MODELOS

Lo mas importante que teneis que saber es:

tipos de modelos de Machine Learning, (los que vamos a ver):

1. Clasificación: En la clasificación, el objetivo es predecir en qué grupo o categoría pertenece algo, como decidir si un correo electrónico es spam o no spam. Algunos ejemplos de algoritmos de clasificación son los árboles de decisión, k-Nearest Neighbors (k-NN) y la regresión logística.

2. Regresión: En la regresión, se intenta predecir un valor numérico, como estimar el precio de una casa basándose en sus características. Ejemplos de algoritmos de regresión incluyen la regresión lineal y las máquinas de vectores de soporte (SVM) para regresión.

3. Clustering (agrupamiento): El clustering se utiliza para agrupar cosas similares juntas sin conocer las categorías de antemano. Por ejemplo, agrupar clientes en función de sus hábitos de compra. Algunos ejemplos de algoritmos de clustering son K-means y DBSCAN.

4. Asociación: La asociación busca patrones o relaciones entre cosas, como descubrir qué productos se compran juntos con frecuencia en una tienda. Un ejemplo de algoritmo de asociación es el algoritmo Apriori.

5. Reinforcement Learning (aprendizaje por refuerzo): En el aprendizaje por refuerzo, un agente (como un robot o un programa) aprende a tomar decisiones para obtener la mayor recompensa posible a través de la experiencia y la prueba y error. Este tipo de aprendizaje se utiliza en situaciones donde no hay una respuesta única y el agente debe aprender a adaptarse a diferentes situaciones. Ejemplos de aplicaciones incluyen enseñar a un programa a jugar a videojuegos o entrenar a un robot para moverse en un entorno desconocido.

##CLASIFICACIÓN

### k-Nearest Neighbors (k-NN)

ejemplo sencillo de un modelo de clasificación utilizando el conjunto de datos Iris y el algoritmo de clasificación k-Nearest Neighbors (k-NN) de la biblioteca scikit-learn.

El conjunto de datos Iris contiene 150 instancias de flores Iris, cada una con cuatro características (longitud del sépalo, anchura del sépalo, longitud del pétalo y anchura del pétalo). El objetivo es clasificar las flores en una de las tres especies posibles (Iris setosa, Iris versicolor o Iris virginica).

Explicación:

1. Importamos las bibliotecas necesarias, incluidas las funciones para cargar el conjunto de datos, dividir los datos, escalar las características, crear el clasificador k-NN y calcular métricas de rendimiento.

2. Cargamos el conjunto de datos Iris y dividimos los datos en conjuntos de entrenamiento (80%) y prueba (20%).

3. Utilizamos el StandardScaler para escalar las características. Esto es importante, ya que el algoritmo k-NN es sensible a la escala de las características.

4. Creamos un clasificador k-NN con n_neighbors=3, lo que significa que el modelo considerará los 3 vecinos más cercanos para tomar una decisión de clasificación. Entrenamos el modelo utilizando el conjunto de entrenamiento escalado.

5. Realizamos predicciones en el conjunto de prueba escalado y evaluamos el rendimiento del modelo calculando la precisión (accuracy), la matriz de confusión y el informe de clasificación. La precisión es la proporción de predicciones correctas entre todas las predicciones realizadas, la matriz de confusión muestra el número de predicciones correctas e incorrectas para cada clase y el informe de clasificación proporciona métricas adicionales como precisión, exhaustividad (recall) y puntuación F1.

Después de ejecutar este código, obtendrás una estimación del rendimiento del modelo de clasificación k

1. Importa las bibliotecas necesarias:

In [ ]:
import numpy as np
import pandas as pd
from sklearn import datasets
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

2. Carga el conjunto de datos Iris y divide los datos en conjuntos de entrenamiento y prueba:

In [ ]:
iris = datasets.load_iris()
X = iris.data
y = iris.target

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

3. Escala las características utilizando el StandardScaler:

In [ ]:
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

4. Crea y entrena el modelo k-NN:

In [ ]:
knn = KNeighborsClassifier(n_neighbors=3)
knn.fit(X_train_scaled, y_train)

KNeighborsClassifier(n_neighbors=3)

5. Realiza predicciones en el conjunto de prueba y evalúa el rendimiento del modelo:

In [ ]:
y_pred = knn.predict(X_test_scaled)
accuracy = accuracy_score(y_test, y_pred)
confusion_mtx = confusion_matrix(y_test, y_pred)
report = classification_report(y_test, y_pred)

print("Accuracy:", accuracy)
print("Confusion Matrix:\n", confusion_mtx)
print("Classification Report:\n", report)

Accuracy: 1.0
Confusion Matrix:
 [[10  0  0]
 [ 0  9  0]
 [ 0  0 11]]
Classification Report:
               precision    recall  f1-score   support

           0       1.00      1.00      1.00        10
           1       1.00      1.00      1.00         9
           2       1.00      1.00      1.00        11

    accuracy                           1.00        30
   macro avg       1.00      1.00      1.00        30
weighted avg       1.00      1.00      1.00        30



Estos resultados indican que el modelo de clasificación k-NN ha realizado un trabajo excelente en la predicción de las especies de flores Iris en el conjunto de datos de prueba. A continuación, se explica cada resultado:

1. Accuracy (Precisión): La precisión es la proporción de predicciones correctas entre todas las predicciones realizadas. En este caso, la precisión es 1.0, lo que indica que el modelo ha clasificado correctamente el 100% de las instancias en el conjunto de datos de prueba.

2. Confusion Matrix (Matriz de confusión): La matriz de confusión muestra el número de predicciones correctas e incorrectas para cada clase. En este caso, se puede ver que el modelo ha clasificado correctamente todas las instancias de cada clase (10 de la clase 0, 9 de la clase 1 y 11 de la clase 2) sin errores.

3. Classification Report (Informe de clasificación): El informe de clasificación proporciona métricas adicionales para evaluar el rendimiento del modelo, como precisión (no confundir con "accuracy"), exhaustividad (recall) y puntuación F1. Estas métricas se calculan por separado para cada clase y se presentan en promedios ponderados y no ponderados.

En este caso, todas las métricas son 1.0 para cada clase, lo que indica un rendimiento perfecto del modelo en el conjunto de datos de prueba. La precisión mide la proporción de predicciones correctas entre todas las predicciones realizadas para una clase específica, la exhaustividad (recall) mide la proporción de instancias correctamente clasificadas de una clase específica entre todas las instancias de esa clase, y la puntuación F1 es una medida armónica entre precisión y exhaustividad (recall) que busca un equilibrio entre ambas métricas.

### REGRESIÓN LOGÍSTICA

El proceso es muy similar al ejemplo anterior de k-NN:

In [ ]:
import numpy as np
import pandas as pd
from sklearn import datasets
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

# Cargamos el conjunto de datos Iris
iris = datasets.load_iris()
X = iris.data
y = iris.target

# Dividimos los datos en conjuntos de entrenamiento (80%) y prueba (20%)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Utilizamos el StandardScaler para escalar las características
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Creamos un clasificador de regresión logística
logreg = LogisticRegression()
logreg.fit(X_train_scaled, y_train)

# Realizamos predicciones en el conjunto de prueba escalado
y_pred = logreg.predict(X_test_scaled)

# Calculamos las métricas de rendimiento
accuracy = accuracy_score(y_test, y_pred)
confusion_mtx = confusion_matrix(y_test, y_pred)
report = classification_report(y_test, y_pred)

print("Accuracy:", accuracy)
print("Confusion Matrix:\n", confusion_mtx)
print("Classification Report:\n", report)


### MAQUINAS DE SOPORTE VECTORIAL (SVM)

In [ ]:
import numpy as np
import pandas as pd
from sklearn import datasets
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

# Cargamos el conjunto de datos Iris
iris = datasets.load_iris()
X = iris.data
y = iris.target

# Dividimos los datos en conjuntos de entrenamiento (80%) y prueba (20%)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Utilizamos el StandardScaler para escalar las características
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Creamos una Máquina de Vectores de Soporte
svm = SVC(kernel="rbf")
svm.fit(X_train_scaled, y_train)

# Realizamos predicciones en el conjunto de prueba escalado
y_pred = svm.predict(X_test_scaled)

# Calculamos las métricas de rendimiento
accuracy = accuracy_score(y_test, y_pred)
confusion_mtx = confusion_matrix(y_test, y_pred)
report = classification_report(y_test, y_pred)

print("Accuracy:", accuracy)
print("Confusion Matrix:\n", confusion_mtx)
print("Classification Report:\n", report)


Accuracy: 1.0
Confusion Matrix:
 [[10  0  0]
 [ 0  9  0]
 [ 0  0 11]]
Classification Report:
               precision    recall  f1-score   support

           0       1.00      1.00      1.00        10
           1       1.00      1.00      1.00         9
           2       1.00      1.00      1.00        11

    accuracy                           1.00        30
   macro avg       1.00      1.00      1.00        30
weighted avg       1.00      1.00      1.00        30



- Este código sigue los mismos pasos que los ejemplos anteriores, pero utiliza un clasificador de Máquinas de Vectores de Soporte (SVM) en lugar de k-NN o Regresión Logística.

- Una SVM es un tipo de algoritmo de aprendizaje automático utilizado para la clasificación y la regresión. En el contexto de la clasificación, como en este caso, una SVM intenta encontrar el hiperplano que maximiza el margen entre las diferentes clases en el espacio de características. En otras palabras, intenta encontrar la línea o plano que mejor separa las diferentes clases.

- El modelo se entrena en el conjunto de entrenamiento escalado, luego se utilizan las características escaladas del conjunto de prueba para hacer predicciones. Luego se calcula la precisión, la matriz de confusión y el informe de clasificación para evaluar el rendimiento del modelo. La precisión es la proporción de predicciones correctas entre todas las predicciones realizadas, la matriz de confusión muestra el número de predicciones correctas e incorrectas para cada clase, y el informe de clasificación proporciona métricas adicionales como precisión, exhaustividad (recall) y puntuación F1.

### RANDOM FOREST

algoritmo de Bosques Aleatorios (Random Forest) para clasificar el conjunto de datos de Iris.

In [ ]:
import numpy as np
import pandas as pd
from sklearn import datasets
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

# Cargamos el conjunto de datos Iris
iris = datasets.load_iris()
X = iris.data
y = iris.target

# Dividimos los datos en conjuntos de entrenamiento (80%) y prueba (20%)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Utilizamos el StandardScaler para escalar las características
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Creamos un clasificador de Bosques Aleatorios
rfc = RandomForestClassifier(n_estimators=100, random_state=42)
rfc.fit(X_train_scaled, y_train)

# Realizamos predicciones en el conjunto de prueba escalado
y_pred = rfc.predict(X_test_scaled)

# Calculamos las métricas de rendimiento
accuracy = accuracy_score(y_test, y_pred)
confusion_mtx = confusion_matrix(y_test, y_pred)
report = classification_report(y_test, y_pred)

print("Accuracy:", accuracy)
print("Confusion Matrix:\n", confusion_mtx)
print("Classification Report:\n", report)


Accuracy: 1.0
Confusion Matrix:
 [[10  0  0]
 [ 0  9  0]
 [ 0  0 11]]
Classification Report:
               precision    recall  f1-score   support

           0       1.00      1.00      1.00        10
           1       1.00      1.00      1.00         9
           2       1.00      1.00      1.00        11

    accuracy                           1.00        30
   macro avg       1.00      1.00      1.00        30
weighted avg       1.00      1.00      1.00        30



Este código sigue los mismos pasos que los ejemplos anteriores, pero utiliza un clasificador de Bosques Aleatorios en lugar de SVM, k-NN o Regresión Logística.

Un Bosque Aleatorio es un algoritmo de aprendizaje automático que genera múltiples árboles de decisión durante el entrenamiento y hace la predicción basándose en la votación de múltiples árboles. En este caso, el clasificador de Bosques Aleatorios se crea con 100 árboles de decisión (n_estimators=100).

El modelo se entrena en el conjunto de entrenamiento escalado, luego se utilizan las características escaladas del conjunto de prueba para hacer predicciones. Finalmente, se calcula la precisión, la matriz de confusión y el informe de clasificación para evaluar el rendimiento del modelo.

## REGRESIÓN

Los modelos de regresión son como recetas de cocina para predecir números.

Imagina que estás tratando de predecir cuántos invitados vendrán a tu próxima fiesta. Puedes tener en cuenta cosas como el día de la semana, el clima, cuántos amigos tienes, etc. Un modelo de regresión toma todos estos factores, los mezcla de una manera particular (la "receta") y produce una predicción del número de invitados.

Así que básicamente, si tienes un montón de datos y quieres predecir un número basado en esos datos, usarías un modelo de regresión. Los diferentes "tipos" de modelos de regresión son simplemente diferentes "recetas" que puedes usar, algunas más sencillas y otras más complicadas, dependiendo de lo complicado que sea el problema que estás tratando de resolver.

### MODELO REGRESIÓN LINEAL

Este es el modelo de regresión más simple. Intenta ajustar una línea a los datos que minimiza la suma de los cuadrados de los residuos.

Este conjunto de datos contiene información recogida por el U.S. Census Service sobre viviendas en el área de Boston. Intentaremos predecir el valor mediano de las casas ocupadas por sus propietarios, en miles (la variable 'MEDV').

In [ ]:
from sklearn.datasets import fetch_california_housing
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

# Cargamos el conjunto de datos de California Housing
california = fetch_california_housing()
X = california.data
y = california.target

# Dividimos los datos en conjuntos de entrenamiento (80%) y prueba (20%)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Utilizamos el StandardScaler para escalar las características
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Creamos una Regresión Lineal
lr = LinearRegression()
lr.fit(X_train_scaled, y_train)

# Realizamos predicciones en el conjunto de prueba escalado
y_pred = lr.predict(X_test_scaled)

# Calculamos las métricas de rendimiento
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print("MSE:", mse)
print("R2 score:", r2)



MSE: 0.5558915986952442
R2 score: 0.575787706032451


- **MSE (Mean Squared Error)**: Imagina que eres un arquero. Cada vez que disparas una flecha a un blanco, la distancia entre donde aterrizó tu flecha (la predicción de tu modelo) y el centro del blanco (el valor real) es un "error". Si disparas muchas flechas, algunas aterrizarán más cerca del centro y otras más lejos. El "Mean Squared Error" (Error Cuadrático Medio) es simplemente el promedio de los cuadrados de todas estas distancias. Cuanto más pequeño sea el MSE, más cerca estarán tus flechas (predicciones) del centro del blanco (valores reales), y por lo tanto, mejor será tu modelo.

- **R2 (R-squared)**: Ahora imagina que tienes un amigo que también está disparando flechas al mismo blanco. Pero tu amigo dispara con los ojos cerrados, así que sus flechas aterrizan en cualquier parte, sin ninguna precisión. Esto es lo que llamamos un "modelo base", que es un modelo muy simple que simplemente predice el promedio de todos los valores reales. El R-squared, o coeficiente de determinación, mide cuánto mejor lo hace tu modelo en comparación con este amigo que dispara con los ojos cerrados. Si R2 es 1, eso significa que tu modelo es perfecto. Si R2 es 0, eso significa que tu modelo no es mejor que tu amigo que dispara con los ojos cerrados. Y si R2 es negativo, eso significa que tu modelo es incluso peor que tu amigo que dispara con los ojos cerrados. Por lo tanto, cuanto más cerca esté R2 de 1, mejor será tu modelo

si quereis un resumen rapido de lo que significa IA, es basicamente predecir a eso se encarga un Data scientist a predecir con datos.

- Aquí teneis el ejemplos de eso: (lo que hay que predecir y la predicción)

In [ ]:
y_test, y_pred

(array([0.477  , 0.458  , 5.00001, ..., 5.00001, 0.723  , 1.515  ]),
 array([0.71912284, 1.76401657, 2.70965883, ..., 4.46877017, 1.18751119,
        2.00940251]))

en este caso es bastante mala :), pero con otros modelos la mejoraremos, cada modelo tiene su cosa.

### MODELO REGRESIÓN POLINÓMICA

La regresión polinómica no es un modelo independiente en scikit-learn como lo son la regresión lineal, Lasso o Ridge, pero se puede implementar utilizando la clase PolynomialFeatures en combinación con un modelo de regresión lineal.

In [ ]:
from sklearn.datasets import fetch_california_housing
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, PolynomialFeatures
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.pipeline import make_pipeline

# Cargamos el conjunto de datos de vivienda de California
california = fetch_california_housing()
X = california.data
y = california.target

# Dividimos los datos en conjuntos de entrenamiento (80%) y prueba (20%)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Creamos un pipeline que primero transforma los datos a características polinomiales, luego los escala y finalmente aplica una regresión lineal
model = make_pipeline(PolynomialFeatures(degree=2), StandardScaler(), LinearRegression())

# Entrenamos el modelo
model.fit(X_train, y_train)

# Realizamos predicciones en el conjunto de prueba
y_pred = model.predict(X_test)

# Calculamos las métricas de rendimiento
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print("MSE:", mse)
print("R2 score:", r2)


MSE: 0.4643015238301156
R2 score: 0.6456819729261956


En este ejemplo, primero transformamos las características a un espacio polinomial utilizando PolynomialFeatures(degree=2). Esto significa que nuestro modelo no solo considerará las características originales, sino también sus interacciones y sus cuadrados. Por ejemplo, si nuestras características originales son [a, b], las características polinomiales de grado 2 serían [1, a, b, a^2, ab, b^2].

Luego, escalamos las características utilizando StandardScaler y aplicamos una regresión lineal. Esto es necesario porque la regresión polinómica sigue siendo una regresión lineal desde el punto de vista de los parámetros que estamos estimando, es decir, estamos buscando una combinación lineal de las características y sus transformaciones polinomiales que mejor se ajuste a los datos.

Finalmente, entrenamos el modelo, realizamos las predicciones y calculamos las métricas de rendimiento de la misma manera que en los ejemplos anteriores.

### MODELO REGRESIÓN DE RIDGE

En este código, en lugar de utilizar la Regresión Lineal (LinearRegression), estamos utilizando la Regresión Ridge (Ridge). El hiperparámetro alpha controla el grado de regularización en la Regresión Ridge: un alpha mayor significa más regularización y un alpha menor significa menos regularización.

In [ ]:
from sklearn.datasets import fetch_california_housing
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import Ridge
from sklearn.metrics import mean_squared_error, r2_score

# Cargamos el conjunto de datos de California Housing
california = fetch_california_housing()
X = california.data
y = california.target

# Dividimos los datos en conjuntos de entrenamiento (80%) y prueba (20%)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Utilizamos el StandardScaler para escalar las características
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Creamos una Regresión Ridge
ridge = Ridge(alpha=1.0)
ridge.fit(X_train_scaled, y_train)

# Realizamos predicciones en el conjunto de prueba escalado
y_pred = ridge.predict(X_test_scaled)

# Calculamos las métricas de rendimiento
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print("MSE:", mse)
print("R2 score:", r2)


MSE: 0.5558548589435974
R2 score: 0.575815742891368


### MODELO REGRESIÓN DE LASSO

En este código, en lugar de utilizar la Regresión Ridge (Ridge), estamos utilizando la Regresión Lasso (Lasso). Al igual que en la regresión Ridge, el hiperparámetro alpha controla el grado de regularización en la Regresión Lasso: un alpha mayor significa más regularización y un alpha menor significa menos regularización. La principal diferencia entre Lasso y Ridge es que Lasso puede hacer que algunos de los coeficientes de las características sean exactamente cero, lo que significa que efectivamente excluye esas características del modelo.

In [ ]:
from sklearn.datasets import fetch_california_housing
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import Lasso
from sklearn.metrics import mean_squared_error, r2_score

# Cargamos el conjunto de datos de California Housing
california = fetch_california_housing()
X = california.data
y = california.target

# Dividimos los datos en conjuntos de entrenamiento (80%) y prueba (20%)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Utilizamos el StandardScaler para escalar las características
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Creamos una Regresión Lasso
lasso = Lasso(alpha=1.0)
lasso.fit(X_train_scaled, y_train)

# Realizamos predicciones en el conjunto de prueba escalado
y_pred = lasso.predict(X_test_scaled)

# Calculamos las métricas de rendimiento
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print("MSE:", mse)
print("R2 score:", r2)


MSE: 1.3106960720039365
R2 score: -0.00021908714592466794


### MODELO Regresión de vectores de soporte (SVR)

DEPENDIENDO DEL MODELO TARDARÁN MAS O MENOS DEBIDO A QUE SERÁN MAS O MENOS POTENTES

In [ ]:
from sklearn.datasets import fetch_california_housing
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.svm import SVR
from sklearn.metrics import mean_squared_error, r2_score

# Cargamos el conjunto de datos de vivienda de California
california = fetch_california_housing()
X = california.data
y = california.target

# Dividimos los datos en conjuntos de entrenamiento (80%) y prueba (20%)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Utilizamos el StandardScaler para escalar las características
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Creamos una regresión SVR
svr = SVR(kernel='rbf')
svr.fit(X_train_scaled, y_train)

# Realizamos predicciones en el conjunto de prueba escalado
y_pred = svr.predict(X_test_scaled)

# Calculamos las métricas de rendimiento
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print("MSE:", mse)
print("R2 score:", r2)


MSE: 0.3570026426754463
R2 score: 0.7275639524733044


si os fijais tenemos un error pequeño MSE y nos acercamos mucho a los datos reales R2

Las Máquinas de Vectores de Soporte (SVM) utilizan algo llamado "funciones de kernel" para transformar los datos. Estas funciones son como lentes mágicas que permiten a las SVM ver los datos desde diferentes perspectivas. Aquí están algunas de las más comunes:

1. Lineal: Este es el lente más simple. No cambia los datos, solo los mira tal como son. Es bueno para los datos que ya están separados por una línea.

2. Polinomial: Este lente permite a las SVM ver interacciones entre las características. Es como si pudiera ver cómo cambian las cosas cuando las características trabajan juntas. Es bueno para los datos que tienen relaciones complejas.

3. RBF (Radial Basis Function o Función de Base Radial): Este es un lente muy flexible que puede ajustarse para ver los datos de muchas formas diferentes. Es como si tuviera diferentes niveles de zoom. Es bueno para los datos que tienen grupos o regiones.

4. Sigmoidal: Este lente transforma los datos de una forma similar a la función sigmoide utilizada en la regresión logística y las redes neuronales. Es como si pudiera ver la probabilidad de que algo suceda.

En resumen, cada kernel tiene sus fortalezas y debilidades, y la elección del kernel puede depender mucho del conjunto de datos con el que estés trabajando.

### MODELO DE ARBOLES DE DECISIÓN

En este ejemplo, utilizamos DecisionTreeRegressor de scikit-learn para crear el modelo de regresión. Los árboles de decisión son un tipo de modelo que hace predicciones siguiendo una serie de decisiones basadas en las características de los datos, parecido a un flujo de decisiones.

El parámetro max_depth controla la profundidad máxima del árbol. Los árboles más profundos pueden capturar relaciones más complejas, pero también son más propensos al sobreajuste. En este ejemplo, hemos establecido max_depth=3, pero podrías ajustar este parámetro dependiendo de tu conjunto de datos y el rendimiento del modelo.

Después de entrenar el modelo y hacer predicciones, calculamos el error cuadrado medio (MSE) y el coeficiente de determinación (R^2) para evaluar el rendimiento del modelo.

In [ ]:
from sklearn.datasets import fetch_california_housing
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import mean_squared_error, r2_score

# Cargamos el conjunto de datos de vivienda de California
california = fetch_california_housing()
X = california.data
y = california.target

# Dividimos los datos en conjuntos de entrenamiento (80%) y prueba (20%)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Creamos un modelo de regresión con árbol de decisión
tree = DecisionTreeRegressor(max_depth=10)

# Entrenamos el modelo
tree.fit(X_train, y_train)

# Realizamos predicciones en el conjunto de prueba
y_pred = tree.predict(X_test)

# Calculamos las métricas de rendimiento
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print("MSE:", mse)
print("R2 score:", r2)


MSE: 0.4203616530141936
R2 score: 0.6792133906328317


### MODELO DE BOSQUES ALEATORIOS

En este caso, estamos usando RandomForestRegressor para crear un modelo de regresión con bosques aleatorios. Los bosques aleatorios son un tipo de modelo de aprendizaje automático que crea un conjunto de árboles de decisión y hace predicciones promediando las predicciones de cada árbol.

El parámetro n_estimators controla el número de árboles en el bosque, y max_depth controla la profundidad máxima de los árboles.

Al igual que antes, después de entrenar el modelo y hacer predicciones, calculamos el error cuadrado medio (MSE) y el coeficiente de determinación (R^2) para evaluar el rendimiento del modelo.

In [ ]:
from sklearn.datasets import fetch_california_housing
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error, r2_score

# Cargamos el conjunto de datos de vivienda de California
california = fetch_california_housing()
X = california.data
y = california.target

# Dividimos los datos en conjuntos de entrenamiento (80%) y prueba (20%)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Creamos un modelo de regresión con bosques aleatorios
forest = RandomForestRegressor(n_estimators=300, max_depth=10, random_state=42)

# Entrenamos el modelo
forest.fit(X_train, y_train)

# Realizamos predicciones en el conjunto de prueba
y_pred = forest.predict(X_test)

# Calculamos las métricas de rendimiento
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print("MSE:", mse)
print("R2 score:", r2)


MSE: 0.29499644957639853
R2 score: 0.7748821516986221


Este modelo es muy potente pero tambien en cuanto computación es bastante costoso

##MODELOS DE ASOCIACIÓN

Los modelos de asociación son una técnica fundamental en inteligencia artificial y aprendizaje automático, utilizada para descubrir relaciones o asociaciones entre conjuntos de elementos en grandes bases de datos.

### Algoritmo Apriori

tener instalado la biblioteca apriori

In [ ]:
pip install apyori


Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
  Preparing metadata (setup.py) ... done
  Created wheel for apyori: filename=apyori-1.1.2-py3-none-any.whl size=5956 sha256=c5a427c1edfcc1f6d751edc5d70fae0e8c5482d27ac0eb0fb591a64b477e5383
  Stored in directory: /root/.cache/pip/wheels/c4/1a/79/20f55c470a50bb3702a8cb7c94d8ada15573538c7f4baebe2d
Successfully built apyori


Ahora, vamos a suponer que tenemos un conjunto de datos de transacciones de una tienda de comestibles en una lista de listas. Cada lista representa una transacción y contiene los productos comprados:

In [ ]:
import pandas as pd
from apyori import apriori

# Ejemplo de datos de transacciones de una tienda
transacciones = [
    ['leche', 'huevos', 'pan'],
    ['leche', 'mantequilla', 'pan'],
    ['huevos', 'mantequilla', 'azúcar'],
    ['leche', 'huevos', 'mantequilla', 'pan'],
    ['huevos', 'azúcar'],
    ['leche', 'huevos', 'mantequilla', 'azúcar'],
    ['leche', 'huevos', 'pan', 'azúcar'],
    ['huevos', 'pan'],
]

# Aplicamos el algoritmo Apriori
soporte_minimo = 0.3
confianza_minima = 0.6
longitud_minima = 2

reglas = apriori(transacciones, min_support=soporte_minimo, min_confidence=confianza_minima, min_length=longitud_minima)

# Imprimimos las reglas de asociación descubiertas
for item in reglas:
    pair = item[0]
    items = [x for x in pair]
    if len(items) > 1:
        print("Regla: " + items[0] + " -> " + items[1])
        print("Soporte: " + str(item[1]))
        print("Confianza: " + str(item[2][0][2]))
        print("Lift: " + str(item[2][0][3]))
        print("=====================================")


Regla: huevos -> azúcar
Soporte: 0.5
Confianza: 1.0
Lift: 1.1428571428571428
Regla: huevos -> leche
Soporte: 0.5
Confianza: 0.8
Lift: 0.9142857142857144
Regla: huevos -> mantequilla
Soporte: 0.375
Confianza: 0.75
Lift: 0.8571428571428571
Regla: huevos -> pan
Soporte: 0.5
Confianza: 0.8
Lift: 0.9142857142857144
Regla: leche -> mantequilla
Soporte: 0.375
Confianza: 0.6
Lift: 1.2
Regla: pan -> leche
Soporte: 0.5
Confianza: 0.8
Lift: 1.28
Regla: huevos -> leche
Soporte: 0.375
Confianza: 0.6
Lift: 1.2


###Algoritmo FP-Growth

El algoritmo FP-Growth es otro algoritmo popular para la minería de reglas de asociación que puede ser más rápido que el algoritmo Apriori en muchos casos. Aquí hay un ejemplo de cómo usar el algoritmo FP-Growth en Python con la biblioteca mlxtend.

In [ ]:
pip install --upgrade mlxtend


Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.4/1.4 MB 10.3 MB/s eta 0:00:00
  Attempting uninstall: mlxtend
    Found existing installation: mlxtend 0.14.0
    Uninstalling mlxtend-0.14.0:
      Successfully uninstalled mlxtend-0.14.0


In [ ]:
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import fpgrowth, association_rules
import pandas as pd

# Ejemplo de datos de transacciones de una tienda
transacciones = [
    ['leche', 'huevos', 'pan'],
    ['leche', 'mantequilla', 'pan'],
    ['huevos', 'mantequilla', 'azúcar'],
    ['leche', 'huevos', 'mantequilla', 'pan'],
    ['huevos', 'azúcar'],
    ['leche', 'huevos', 'mantequilla', 'azúcar'],
    ['leche', 'huevos', 'pan', 'azúcar'],
    ['huevos', 'pan'],
]

# Transformamos los datos a un formato adecuado para mlxtend
te = TransactionEncoder()
te_ary = te.fit(transacciones).transform(transacciones)
df = pd.DataFrame(te_ary, columns=te.columns_)

# Aplicamos el algoritmo FP-Growth
frequent_itemsets = fpgrowth(df, min_support=0.3, use_colnames=True)

# Generamos las reglas de asociación
reglas = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.6)

# Imprimimos las reglas
print(reglas)


        antecedents      consequents  antecedent support  consequent support  \
0             (pan)         (huevos)               0.625               0.875   
1             (pan)          (leche)               0.625               0.625   
2           (leche)            (pan)               0.625               0.625   
3           (leche)         (huevos)               0.625               0.875   
4      (pan, leche)         (huevos)               0.500               0.875   
5     (pan, huevos)          (leche)               0.500               0.625   
6   (leche, huevos)            (pan)               0.500               0.625   
7             (pan)  (leche, huevos)               0.625               0.500   
8           (leche)    (pan, huevos)               0.625               0.500   
9           (leche)    (mantequilla)               0.625               0.500   
10    (mantequilla)          (leche)               0.500               0.625   
11    (mantequilla)         (huevos)    

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


###Algoritmo Eclat

El algoritmo Eclat es un método de minería de conjuntos de ítems frecuentes que se centra sólo en el soporte del ítem, a diferencia de Apriori o FP-Growth que también consideran la confianza y el lift para la generación de reglas de asociación.

Actualmente, no existe una implementación del algoritmo Eclat en las bibliotecas más populares de Python como scikit-learn o mlxtend.

In [ ]:
import pandas as pd
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori

# Ejemplo de datos de transacciones de una tienda
transacciones = [
    ['leche', 'huevos', 'pan'],
    ['leche', 'mantequilla', 'pan'],
    ['huevos', 'mantequilla', 'azúcar'],
    ['leche', 'huevos', 'mantequilla', 'pan'],
    ['huevos', 'azúcar'],
    ['leche', 'huevos', 'mantequilla', 'azúcar'],
    ['leche', 'huevos', 'pan', 'azúcar'],
    ['huevos', 'pan'],
]

# Transformamos los datos a un formato adecuado para mlxtend
te = TransactionEncoder()
te_ary = te.fit(transacciones).transform(transacciones)
df = pd.DataFrame(te_ary, columns=te.columns_)

# Aplicamos el algoritmo Apriori para obtener los conjuntos de ítems frecuentes
frequent_itemsets = apriori(df, min_support=0.3, use_colnames=True)

# Solo nos interesan los conjuntos de ítems frecuentes, por lo que no generamos las reglas de asociación

print(frequent_itemsets)


    support               itemsets
0     0.500               (azúcar)
1     0.875               (huevos)
2     0.625                (leche)
3     0.500          (mantequilla)
4     0.625                  (pan)
5     0.500       (azúcar, huevos)
6     0.500        (leche, huevos)
7     0.375  (huevos, mantequilla)
8     0.500          (pan, huevos)
9     0.375   (leche, mantequilla)
10    0.500           (pan, leche)
11    0.375   (pan, leche, huevos)


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


Este código utiliza la función apriori de la biblioteca mlxtend para encontrar los conjuntos de ítems frecuentes con un soporte mínimo del 30%. No genera las reglas de asociación, ya que el algoritmo Eclat no las considera. Al final, se imprimen los conjuntos de ítems frecuentes.

###Algoritmo de reglas de asociación generalizada (GAR)

Las Reglas de Asociación Generalizadas (GAR) son una extensión del concepto de reglas de asociación que permiten utilizar conceptos jerárquicos en los datos. Por ejemplo, en un conjunto de datos de transacciones de una tienda de comestibles, un concepto jerárquico podría ser que "fruta" incluye "manzanas", "plátanos", "naranjas", etc.

Actualmente, las bibliotecas de aprendizaje automático más populares de Python, como scikit-learn o mlxtend, no proporcionan una implementación del algoritmo de Reglas de Asociación Generalizadas. Este tipo de análisis requiere un preprocesamiento considerable de los datos y una definición clara de las jerarquías de los ítems, lo que puede variar mucho dependiendo del conjunto de datos específico.

Además, el algoritmo de Reglas de Asociación Generalizadas puede ser bastante complejo y computacionalmente intensivo, ya que necesita considerar todas las posibles combinaciones de ítems y sus conceptos superiores en la jerarquía.

Si tienes un conjunto de datos con una jerarquía de ítems claramente definida y necesitas aplicar las Reglas de Asociación Generalizadas, podrías necesitar implementar el algoritmo tú mismo o buscar una biblioteca de aprendizaje automático más especializada que lo incluya. Alternativamente, podrías considerar simplificar el problema utilizando reglas de asociación estándar y tratando los conceptos superiores en la jerarquía como ítems separados.

podemos simular un ejemplo simple de GAR a través de la creación manual de jerarquías y la implementación de reglas de asociación. A continuación se muestra un ejemplo de cómo podríamos hacer esto utilizando la biblioteca mlxtend y el algoritmo Apriori:

In [ ]:
import pandas as pd
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori, association_rules

# Ejemplo de datos de transacciones de una tienda
transacciones = [
    ['manzanas', 'huevos', 'pan'],
    ['plátanos', 'mantequilla', 'pan'],
    ['manzanas', 'mantequilla', 'azúcar'],
    ['plátanos', 'huevos', 'mantequilla', 'pan'],
    ['manzanas', 'azúcar'],
    ['plátanos', 'huevos', 'mantequilla', 'azúcar'],
    ['plátanos', 'huevos', 'pan', 'azúcar'],
    ['manzanas', 'pan'],
]

# Transformamos los datos a un formato adecuado para mlxtend
te = TransactionEncoder()
te_ary = te.fit(transacciones).transform(transacciones)
df = pd.DataFrame(te_ary, columns=te.columns_)

# Aplicamos el algoritmo Apriori para obtener los conjuntos de ítems frecuentes
frequent_itemsets = apriori(df, min_support=0.3, use_colnames=True)

# Generamos las reglas de asociación
rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.6)

# Agregamos manualmente la jerarquía, sustituyendo 'manzanas' y 'plátanos' por 'fruta'
rules["antecedents"] = rules["antecedents"].apply(lambda x: x if not ('manzanas' in x or 'plátanos' in x) else frozenset({'fruta'}))
rules["consequents"] = rules["consequents"].apply(lambda x: x if not ('manzanas' in x or 'plátanos' in x) else frozenset({'fruta'}))

print(rules)


     antecedents    consequents  antecedent support  consequent support  \
0          (pan)       (huevos)               0.625               0.500   
1       (huevos)          (pan)               0.500               0.625   
2        (fruta)       (huevos)               0.500               0.500   
3       (huevos)        (fruta)               0.500               0.500   
4        (fruta)  (mantequilla)               0.500               0.500   
5  (mantequilla)        (fruta)               0.500               0.500   
6          (pan)        (fruta)               0.625               0.500   
7        (fruta)          (pan)               0.500               0.625   

   support  confidence  lift  leverage  conviction  zhangs_metric  
0    0.375        0.60   1.2    0.0625        1.25       0.444444  
1    0.375        0.75   1.2    0.0625        1.50       0.333333  
2    0.375        0.75   1.5    0.1250        2.00       0.666667  
3    0.375        0.75   1.5    0.1250        2.00  

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


En este ejemplo, primero generamos las reglas de asociación estándar con el algoritmo Apriori. Luego, reemplazamos manualmente 'manzanas' y 'plátanos' por 'fruta' en las reglas de asociación para simular una jerarquía de ítems.

Por favor ten en cuenta que este es un ejemplo muy simplificado y que la implementación real de las Reglas de Asociación Generalizadas puede ser mucho más compleja, especialmente para conjuntos de datos grandes con jerarquías de ítems más complicadas.

###Algoritmo OPUS AR  (OPtimized Universal Set - Association Rules)

OPUS-AR es un algoritmo de minería de reglas de asociación que utiliza una estrategia de búsqueda de conjunto universal optimizado. Es un algoritmo bastante nuevo y especializado, y no se usa tan comúnmente como Apriori, FP-Growth, o Eclat.

Si tienes un conjunto de datos y necesitas aplicar el algoritmo OPUS-AR, podrías necesitar implementar el algoritmo tú mismo en Python, basándote en las descripciones del algoritmo en la literatura de minería de datos. Alternativamente, podrías buscar una biblioteca de aprendizaje automático más especializada que pueda tener una implementación de OPUS-AR, o podrías considerar utilizar un software de minería de datos más especializado que incluya OPUS-AR.

##MODELOS DE REINFORCEMENT LEARNING

El Aprendizaje por Reforzamiento (Reinforcement Learning, RL) es una rama del Aprendizaje Automático (Machine Learning) que se centra en cómo un agente puede aprender a tomar acciones en un entorno para maximizar algún tipo de recompensa acumulativa. Existen varias arquitecturas y algoritmos en RL, cada uno con diferentes enfoques para resolver el problema del aprendizaje por refuerzo.

-EJEMPLOS:

### Q-Learning:

En Q-Learning, el agente aprende una función de valor llamada Q, que da la recompensa esperada para cada acción en cada estado. El agente luego utiliza esta función para elegir las acciones que maximizan la recompensa esperada.

La idea básica detrás de Q-learning es que, para cada par estado-acción, se actualiza el valor Q utilizando la ecuación de Bellman, que dice que el valor Q de un par estado-acción es igual a la recompensa inmediata más el valor Q descontado del mejor par estado-acción en el siguiente estado.

La ecuación de actualización de Q-learning es la siguiente:

Q(s, a) = Q(s, a) + α * [R(s, a) + γ * max Q(s', a') - Q(s, a)]

Donde:

- s es el estado actual.
- a es la acción tomada.
- α es la tasa de aprendizaje.
- R(s, a) es la recompensa inmediata al tomar la acción a en el estado s.
- γ es el factor de descuento.
- max Q(s', a') es el valor máximo de Q para todas las posibles acciones a' en el siguiente estado s'.
- Q(s, a) es el valor antiguo, y Q(s, a) en el lado izquierdo de la ecuación es el valor nuevo.

Utilizando el entorno "FrozenLake-v1" de OpenAI Gym, que es un laberinto simple donde el objetivo es llegar de la casilla de inicio a la casilla de meta sin caer en los agujeros:

In [ ]:
import numpy as np
import gym

# Crea el entorno
env = gym.make('FrozenLake-v1')

# Inicializa la tabla Q con ceros
Q = np.zeros([env.observation_space.n, env.action_space.n])

# Parámetros
alpha = 0.5
gamma = 0.95
num_episodes = 5000

# Para cada episodio
for i in range(num_episodes):
    # Reinicia el estado
    s = env.reset()
    done = False
    while not done:
        # Elige la acción con el mayor valor Q (añade un poco de ruido para la exploración)
        a = np.argmax(Q[s, :] + np.random.randn(1, env.action_space.n) * (1. / (i + 1)))
        # Realiza la acción
        s1, r, done, _ = env.step(a)
        # Actualiza el valor Q
        Q[s, a] = Q[s, a] + alpha * (r + gamma * np.max(Q[s1, :]) - Q[s, a])
        s = s1

print("Tabla Q final:")
print(Q)


/usr/local/lib/python3.10/dist-packages/gym/core.py:317: DeprecationWarning: WARN: Initializing wrapper in old step API which returns one bool instead of two. It is recommended to set `new_step_api=True` to use new step API. This will be the default behaviour in future.
  deprecation(
/usr/local/lib/python3.10/dist-packages/gym/wrappers/step_api_compatibility.py:39: DeprecationWarning: WARN: Initializing environment in old step API which returns one bool instead of two. It is recommended to set `new_step_api=True` to use new step API. This will be the default behaviour in future.
  deprecation(


Tabla Q final:
[[1.44985075e-01 1.29029475e-02 8.46854361e-03 6.81472763e-03]
 [1.39657126e-03 5.62668262e-03 8.55552670e-04 1.25802125e-03]
 [1.55506335e-02 3.96537224e-03 4.63632381e-03 3.38336485e-03]
 [2.77211155e-03 1.86309367e-03 2.45213247e-03 5.32715045e-03]
 [1.44897663e-01 1.08260392e-04 3.88364481e-03 5.06491503e-04]
 [0.00000000e+00 0.00000000e+00 0.00000000e+00 0.00000000e+00]
 [5.83667464e-04 2.97179075e-04 8.01246235e-03 2.68667095e-04]
 [0.00000000e+00 0.00000000e+00 0.00000000e+00 0.00000000e+00]
 [4.93624559e-03 5.42498852e-03 0.00000000e+00 1.92892990e-01]
 [0.00000000e+00 3.55596793e-01 0.00000000e+00 4.08946251e-03]
 [5.25443719e-01 0.00000000e+00 0.00000000e+00 0.00000000e+00]
 [0.00000000e+00 0.00000000e+00 0.00000000e+00 0.00000000e+00]
 [0.00000000e+00 0.00000000e+00 0.00000000e+00 0.00000000e+00]
 [9.31408627e-04 9.31408627e-04 5.54147302e-01 1.96086027e-03]
 [0.00000000e+00 0.00000000e+00 0.00000000e+00 8.00729505e-01]
 [0.00000000e+00 0.00000000e+00 0.000000

En este código, el agente explora el entorno "FrozenLake-v1" durante un número determinado de episodios. En cada paso, elige la acción con el mayor valor Q en el estado actual, pero también añade un poco de ruido a los valores Q para fomentar la exploración de acciones que no ha probado antes. Luego, realiza la acción, observa la recompensa y el próximo estado, y actualiza el valor Q de la acción que acaba de tomar utilizando la ecuación de actualización de Q-learning.

Al final, imprime la tabla Q final, que contiene los valores Q de todas las acciones en todos los estados. Estos valores representan la recompensa esperada de tomar cada acción en cada estado, y el agente puede usarlos para decidir qué acción tomar en cada estado para maximizar la recompensa total.

### Deep Q-Network (DQN):

DQN es una extensión de Q-learning que utiliza redes neuronales profundas para aproximar la función Q. Esto permite que el algoritmo maneje problemas con grandes espacios de estado que no pueden ser manejados eficazmente con métodos tabulares.

In [ ]:
import gym
import numpy as np
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.optimizers import Adam
from collections import deque
import random

# Parámetros
gamma = 0.95
learning_rate = 0.001
epsilon = 1.0
epsilon_min = 0.01
epsilon_decay = 0.995
batch_size = 64
memory = deque(maxlen=2000)

# Crea el entorno
env = gym.make('CartPole-v1')
state_size = env.observation_space.shape[0]
action_size = env.action_space.n

# Crea la red neuronal
model = Sequential()
model.add(Dense(24, input_dim=state_size, activation='relu'))
model.add(Dense(24, activation='relu'))
model.add(Dense(action_size, activation='linear'))
model.compile(loss='mse', optimizer=Adam(learning_rate=learning_rate))

# Para cada episodio
for i in range(100):
    # Reinicia el estado
    state = env.reset()
    state = np.reshape(state, [1, state_size])
    done = False
    while not done:
        # Elige la acción
        if np.random.rand() <= epsilon:
            action = random.randrange(action_size)
        else:
            action = np.argmax(model.predict(state))
        # Realiza la acción
        next_state, reward, done, _ = env.step(action)
        next_state = np.reshape(next_state, [1, state_size])
        # Almacena la transición
        memory.append((state, action, reward, next_state, done))
        state = next_state
        # Si el episodio ha terminado, entrena la red neuronal con un lote de transiciones
        if done:
            if len(memory) > batch_size:
                minibatch = random.sample(memory, batch_size)
                for state, action, reward, next_state, done in minibatch:
                    target = reward
                    if not done:
                        target += gamma * np.amax(model.predict(next_state)[0])
                    target_f = model.predict(state)
                    target_f[0][action] = target
                    model.fit(state, target_f, epochs=4, verbose=0)
            if epsilon > epsilon_min:
                epsilon *= epsilon_decay

model.save('dqn_model.h5')


### Policy Gradient Methods:

En lugar de aprender una función de valor, los métodos de gradiente de política aprenden directamente una política de acción. Esto se logra maximizando una medida de recompensa acumulativa con respecto a los parámetros de la política. Algunos ejemplos de estos métodos incluyen REINFORCE, Actor-Critic y Proximal Policy Optimization (PPO).

In [ ]:
import gym
import numpy as np
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.optimizers import Adam

# Crea el entorno
env = gym.make('CartPole-v1')
state_size = env.observation_space.shape[0]
action_size = env.action_space.n

# Crea la red neuronal
model = Sequential()
model.add(Dense(24, input_dim=state_size, activation='relu'))
model.add(Dense(24, activation='relu'))
model.add(Dense(action_size, activation='softmax'))
model.compile(loss='categorical_crossentropy', optimizer=Adam())

# Parámetros
gamma = 0.95
num_episodes = 100

# Para cada episodio
for i in range(num_episodes):
    # Reinicia el estado
    state = env.reset()
    state = np.reshape(state, [1, state_size])
    reward_sum = 0
    states = []
    actions = []
    rewards = []
    done = False
    while not done:
        # Elige la acción
        action_prob = model.predict(state)[0]
        action = np.random.choice(action_size, p=action_prob)
        # Realiza la acción
        next_state, reward, done, _ = env.step(action)
        next_state = np.reshape(next_state, [1, state_size])
        # Almacena la transición
        states.append(state[0])
        action_one_hot = np.zeros(action_size)
        action_one_hot[action] = 1
        actions.append(action_one_hot)
        rewards.append(reward)
        state = next_state
        reward_sum += reward
        # Si el episodio ha terminado, actualiza los pesos de la red neuronal
        if done:
            discounted_rewards = np.zeros_like(rewards)
            running_add = 0
            for t in reversed(range(len(rewards))):
                running_add = running_add * gamma + rewards[t]
                discounted_rewards[t] = running_add
            discounted_rewards -= np.mean(discounted_rewards)
            discounted_rewards /= np.std(discounted_rewards)
            states = np.vstack(states)
            actions = np.vstack(actions)
            discounted_rewards = np.vstack(discounted_rewards)
            model.train_on_batch(states, actions, sample_weight=discounted_rewards.ravel())
            states = []
            actions = []
            rewards = []
            print('Episode: %d/1000, Reward: %d' % (i+1, reward_sum))

model.save('pg_model.h5')


### Actor-Critic Methods:

Los métodos Actor-Critic combinan las ideas de los métodos de gradiente de política y los métodos de función de valor. El "actor" aprende la política de acción, mientras que el "crítico" aprende a evaluar la política del actor. Ejemplos de estos métodos incluyen Advantage Actor-Critic (A2C), Asynchronous Advantage Actor-Critic (A3C) y Soft Actor-Critic (SAC).

In [ ]:
import gym
import numpy as np
import tensorflow as tf
from tensorflow.keras.models import Model
from tensorflow.keras.layers import Dense, Input

# Crea el entorno
env = gym.make('CartPole-v1')
state_size = env.observation_space.shape[0]
action_size = env.action_space.n

# Crea la red neuronal del actor
input_state = Input(shape=(state_size,))
dense1 = Dense(24, activation='relu')(input_state)
dense2 = Dense(24, activation='relu')(dense1)
output_probs = Dense(action_size, activation='softmax')(dense2)
actor = Model(inputs=input_state, outputs=output_probs)
actor.compile(loss='categorical_crossentropy', optimizer='adam')

# Crea la red neuronal del crítico
input_state = Input(shape=(state_size,))
dense1 = Dense(24, activation='relu')(input_state)
dense2 = Dense(24, activation='relu')(dense1)
output_value = Dense(1, activation='linear')(dense2)
critic = Model(inputs=input_state, outputs=output_value)
critic.compile(loss='mse', optimizer='adam')

# Parámetros
gamma = 0.95
num_episodes = 1000

# Para cada episodio
for i in range(num_episodes):
    # Reinicia el estado
    state = env.reset()
    state = np.reshape(state, [1, state_size])
    reward_sum = 0
    done = False
    while not done:
        # Elige la acción
        action_prob = actor.predict(state)[0]
        action = np.random.choice(action_size, p=action_prob)
        # Realiza la acción
        next_state, reward, done, _ = env.step(action)
        next_state = np.reshape(next_state, [1, state_size])
        # Calcula la ventaja
        target = reward + (1 - done) * gamma * critic.predict(next_state)[0]
        advantage = target - critic.predict(state)
        # Actualiza el actor
        action_one_hot = np.zeros(action_size)
        action_one_hot[action] = 1
        actor.fit(state, action_one_hot.reshape([1, -1]), sample_weight=advantage, epochs=1, verbose=0)
        # Actualiza el crítico
        critic.fit(state, target, epochs=1, verbose=0)
        state = next_state
        reward_sum += reward
    print('Episode: %d/1000, Reward: %d' % (i+1, reward_sum))

actor.save('actor_model.h5')
critic.save('critic_model.h5')


### Model-Based RL:

En el aprendizaje por refuerzo basado en modelos, el agente aprende un modelo del entorno además de una política o una función de valor. Este modelo se utiliza para predecir los resultados de las acciones, lo que puede mejorar la eficiencia del aprendizaje al permitir la planificación y la simulación.

In [ ]:
import gym
import numpy as np
from sklearn.ensemble import ExtraTreesRegressor
from scipy.optimize import minimize

# Crea el entorno
env = gym.make('MountainCar-v0')

# Inicializa el modelo del entorno con un regresor de árboles extremadamente aleatorios
model = ExtraTreesRegressor()

# Inicializa el conjunto de datos de entrenamiento
X_train = np.empty((0, env.observation_space.shape[0] + 1))
y_train = np.empty((0, env.observation_space.shape[0]))

# Parámetros
num_episodes = 1000
num_steps = 200
num_random_episodes = 100

# Para cada episodio
for i in range(num_episodes):
    # Reinicia el estado
    state = env.reset()
    for _ in range(num_steps):
        # Elige la acción
        if i < num_random_episodes or np.random.rand() < 0.1:
            # Durante los primeros episodios, o con una probabilidad del 10%, elige una acción aleatoria
            action = env.action_space.sample()
        else:
            # Después de eso, elige la acción que minimiza la predicción del modelo del próximo estado
            action = int(minimize(lambda a: model.predict(np.append(state, a).reshape(1, -1)), x0=1, bounds=[(0, env.action_space.n-1)]).x)
        # Realiza la acción
        next_state, reward, done, _ = env.step(action)
        # Almacena la transición en el conjunto de datos de entrenamiento
        X_train = np.vstack([X_train, np.append(state, action)])
        y_train = np.vstack([y_train, next_state])
        state = next_state
        # Si el episodio ha terminado, entrena el modelo del entorno
        if done:
            model.fit(X_train, y_train)
            break


### Inverse Reinforcement Learning (IRL):

En IRL, el objetivo es aprender la función de recompensa que un agente está optimizando a partir de sus acciones observadas, en lugar de aprender una política o una función de valor.

Un algoritmo popular para IRL es el algoritmo de máxima entropía.

In [ ]:
import gym
import numpy as np
from sklearn.linear_model import LinearRegression

# Crea el entorno
env = gym.make('MountainCar-v0')

# Inicializa la función de recompensa con una regresión lineal
reward_function = LinearRegression()

# Inicializa el conjunto de datos de entrenamiento
X_train = np.empty((0, env.observation_space.shape[0]))
y_train = np.empty((0,))

# Parámetros
num_episodes = 1000
num_steps = 200

# Para cada episodio
for i in range(num_episodes):
    # Reinicia el estado
    state = env.reset()
    for _ in range(num_steps):
        # Observa la acción del agente experto
        action = expert_policy(state)
        # Realiza la acción
        next_state, _, done, _ = env.step(action)
        # Almacena la transición en el conjunto de datos de entrenamiento
        X_train = np.vstack([X_train, state])
        y_train = np.append(y_train, expert_reward(state, action))
        state = next_state
        # Si el episodio ha terminado, entrena la función de recompensa
        if done:
            reward_function.fit(X_train, y_train)
            break


En este código, se asume que tienes una política de experto (expert_policy) y una recompensa de experto (expert_reward) que puedes usar para generar los datos de entrenamiento. El agente observa las acciones del experto y las recompensas correspondientes, y luego entrena una función de recompensa para predecir la recompensa del experto a partir del estado.

Por favor, ten en cuenta que este es un ejemplo muy básico de IRL y puede que no funcione bien en entornos más complejos. Para entornos más complejos, podrías necesitar técnicas más avanzadas, como un modelo más preciso de la función de recompensa, una política de experto más sofisticada, o técnicas como el algoritmo de máxima entropía para IRL. Además, este código asume que tienes acceso a una política de experto y a una recompensa de experto, lo cual puede no ser el caso en muchas situaciones reales.

### Monte Carlo Tree Search (MCTS):

MCTS es una técnica de búsqueda que utiliza simulaciones de Monte Carlo para evaluar las acciones. Ha demostrado ser muy eficaz en una serie de tareas, incluyendo juegos como Go y ajedrez.

Aquí tienes un ejemplo simplificado de cómo podría verse un algoritmo de MCTS en Python para el juego de Tic-Tac-Toe:

In [ ]:
import numpy as np
from copy import deepcopy

class Node:
    def __init__(self, state, parent=None):
        self.state = state
        self.parent = parent
        self.children = []
        self.wins = 0
        self.visits = 0

def get_possible_actions(state):
    return [(i, j) for i in range(3) for j in range(3) if state[i][j] == 0]

def get_next_state(state, action, player):
    next_state = deepcopy(state)
    next_state[action[0]][action[1]] = player
    return next_state

def get_winner(state):
    lines = [[state[i][j] for j in range(3)] for i in range(3)] + \
            [[state[j][i] for j in range(3)] for i in range(3)] + \
            [[state[i][i] for i in range(3)]] + \
            [[state[i][2 - i] for i in range(3)]]
    if [1, 1, 1] in lines:
        return 1
    if [-1, -1, -1] in lines:
        return -1
    return 0

def simulate(node):
    state = node.state
    player = -1 if np.sum(state) == 1 else 1
    while True:
        winner = get_winner(state)
        if winner != 0 or np.all(state != 0):
            return winner
        action = np.random.choice(get_possible_actions(state))
        state = get_next_state(state, action, player)
        player *= -1

def update(node, winner):
    while node is not None:
        node.visits += 1
        if winner == 1:
            node.wins += 1
        elif winner == -1:
            node.wins -= 1
        node = node.parent

def mcts(state, num_simulations):
    root = Node(state)
    for _ in range(num_simulations):
        node = root
        while len(get_possible_actions(node.state)) == 0 or len(node.children) > 0:
            if len(node.children) == 0:
                break
            ucb_values = [child.wins / child.visits + np.sqrt(2 * np.log(node.visits) / child.visits) for child in node.children]
            node = node.children[np.argmax(ucb_values)]
        for action in get_possible_actions(node.state):
            node.children.append(Node(get_next_state(node.state, action, 1), node))
        winner = simulate(node)
        update(node, winner)
    return root.children[np.argmax([child.visits for child in root.children])].state


Este código implementa un algoritmo de MCTS para el juego de Tic-Tac-Toe. En cada paso, el algoritmo elige la acción con el mayor valor de UCB (Upper Confidence Bound), realiza la acción, y luego realiza una serie de simulaciones para estimar el valor de la acción. Después de un número determinado de simulaciones, el algoritmo elige la acción

Monte Carlo Tree Search (MCTS) es un algoritmo de búsqueda que se utiliza comúnmente en problemas de toma de decisiones, especialmente en juegos de tablero como Go y ajedrez. MCTS utiliza simulaciones de Monte Carlo para estimar el valor de cada acción, y luego elige la acción con el mayor valor estimado.

##MODELOS AVANZADOS (EN ESTUDIO)

A partir de aquí es un paso mas allá de lo necesario son modelos experimentales y que están en desarrollo

### ARQUITECTURA RED NEURONAL CUÁNTICA

In [ ]:
from qiskit import Aer
from qiskit.circuit.library import ZZFeatureMap
from qiskit_machine_learning.neural_networks import CircuitQNN
from qiskit_machine_learning.connectors import TorchConnector

# Definir el mapa de características cuánticas
feature_map = ZZFeatureMap(feature_dimension=2, reps=2, entanglement='linear')

# Definir la función de interpretación
def interpret(result):
    # Obtén la distribución de probabilidad de la salida
    probabilities = result.get_counts()

    # Calcula y devuelve la paridad
    return [probabilities.get('0', 0) - probabilities.get('1', 0)]

# Definir la red neuronal cuántica
circuit_qnn = CircuitQNN(circuit=feature_map,
                         input_params=feature_map.parameters,
                         interpret=interpret,
                         output_shape=(1,),  # Clasificación binaria
                         quantum_instance=Aer.get_backend('statevector_simulator'))

# Conectar con PyTorch
quantum_model = TorchConnector(circuit_qnn)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


ModuleNotFoundError: ignored

#### EJEMPLO CON DATASET REAL

In [ ]:
!pip install qiskit
!pip install qiskit_machine_learning

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Installing backend dependencies ... done
  Preparing metadata (pyproject.toml) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.9/5.9 MB 35.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.8/12.8 MB 23.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 241.5/241.5 kB 15.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 129.9/129.9 kB 12.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.9/1.9 MB 38.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 49.6/49.6 kB 4.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 110.5/110.5 kB 5.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 49.6/49.6 kB 4.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 37.5/37.5 MB 10.1 MB/s eta 0:00:00


In [ ]:
from qiskit import Aer
from qiskit.circuit.library import ZZFeatureMap
from qiskit_machine_learning.neural_networks import TwoLayerQNN
from qiskit.algorithms.optimizers import COBYLA
from qiskit_machine_learning.algorithms.classifiers import NeuralNetworkClassifier

from sklearn import datasets
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score

# Cargar el conjunto de datos de Iris
iris = datasets.load_iris()
X = iris.data[:100, :2]  # Tomamos solo las primeras dos características de las primeras 100 muestras
y = iris.target[:100]  # Solo las clases 0 y 1

# Normalizar los datos
scaler = StandardScaler().fit(X)
X = scaler.transform(X)

# Dividir los datos en conjuntos de entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Crear la red neuronal cuántica
feature_map = ZZFeatureMap(feature_dimension=2, reps=2)
qnn = TwoLayerQNN(2, feature_map=feature_map, quantum_instance=Aer.get_backend('statevector_simulator'))

# Crear el clasificador de red neuronal cuántica
classifier = NeuralNetworkClassifier(qnn, optimizer=COBYLA())

# Entrenar el clasificador
classifier.fit(X_train, y_train)

# Hacer predicciones
y_pred = classifier.predict(X_test)

# Calcular la precisión del modelo
accuracy = accuracy_score(y_test, y_pred)
print(f'Accuracy: {accuracy * 100}%')


Se han truncado las últimas 5000 líneas del flujo de salida.
/usr/local/lib/python3.10/dist-packages/qiskit/opflow/list_ops/list_op.py:240: DeprecationWarning: The class ``qiskit.opflow.list_ops.composed_op.ComposedOp`` is deprecated as of qiskit-terra 0.24.0. It will be removed no earlier than 3 months after the release date. For code migration guidelines, visit https://qisk.it/opflow_migration.
  return self.__class__(
/usr/local/lib/python3.10/dist-packages/qiskit/opflow/list_ops/composed_op.py:53: DeprecationWarning: The class ``qiskit.opflow.list_ops.list_op.ListOp`` is deprecated as of qiskit-terra 0.24.0. It will be removed no earlier than 3 months after the release date. For code migration guidelines, visit https://qisk.it/opflow_migration.
  super().__init__(oplist, combo_fn=partial(reduce, np.dot), coeff=coeff, abelian=abelian)
/usr/local/lib/python3.10/dist-packages/qiskit/opflow/list_ops/list_op.py:80: DeprecationWarning: The class ``qiskit.opflow.operator_base.OperatorBase

KeyboardInterrupt: ignored

Este es un código que utiliza el aprendizaje automático cuántico (QML, por sus siglas en inglés) para clasificar el conjunto de datos de Iris utilizando el paquete Qiskit de IBM:

1. Importaciones:

Primero, importamos todas las bibliotecas necesarias.

- qiskit: un paquete de IBM para la computación cuántica.
- sklearn: una biblioteca de aprendizaje automático de propósito general.

2. Cargar y preparar el conjunto de datos:

- Usamos datasets.load_iris() para cargar el conjunto de datos de Iris, que es un conjunto de datos comúnmente utilizado en aprendizaje automático y estadísticas. Consiste en 150 muestras de las medidas de las flores de iris de tres especies diferentes.

- Tomamos solo las primeras dos características de las primeras 100 muestras de los datos de entrada X y las correspondientes etiquetas de clase y. Esto se hace para simplificar el problema a un problema de clasificación binaria (en lugar de multiclase) y reducir la dimensionalidad de los datos.

- Normalizamos los datos con StandardScaler().fit(X), que elimina la media y escala los datos a la varianza de la unidad.

3. Dividir los datos:

- Usamos train_test_split para dividir los datos en conjuntos de entrenamiento y prueba, con el 80% de los datos utilizados para el entrenamiento y el 20% para las pruebas.

4. Crear la red neuronal cuántica:

- Creamos un mapa de características cuánticas (ZZFeatureMap) que se utiliza para codificar los datos clásicos en un estado cuántico.

- Luego, creamos una red neuronal cuántica (TwoLayerQNN) con este mapa de características.

5. Crear el clasificador:

- Utilizamos el optimizador COBYLA (Constrained Optimization BY Linear Approximations) para entrenar nuestra red neuronal cuántica.

- Luego, creamos un clasificador (NeuralNetworkClassifier) con la red neuronal cuántica y el optimizador.

6. Entrenamiento del clasificador:

- Usamos classifier.fit(X_train, y_train) para entrenar el clasificador en el conjunto de entrenamiento.

7. Hacer predicciones:

- Usamos classifier.predict(X_test) para hacer predicciones en el conjunto de prueba.

8. Evaluar el modelo:

- Finalmente, calculamos la precisión de nuestras predicciones utilizando accuracy_score(y_test, y_pred) y la imprimimos.

### Redes Neuronales de Espolones Temporales (SNN):

Las SNN son un tipo de red neuronal que intenta imitar más de cerca la forma en que las neuronas reales en el cerebro interactúan. En lugar de tener activaciones continuas, las SNN funcionan con "espolones" discretos en el tiempo, de forma similar a cómo las neuronas biológicas disparan impulsos eléctricos.

### Redes Neuronales Capsulares

Propuestas por Geoffrey Hinton, una de las figuras principales en el aprendizaje profundo, las redes neuronales capsulares intentan solucionar algunos de los problemas con las redes neuronales convolucionales, especialmente en lo que respecta a la detección de jerarquías de características y la invarianza de posición y orientación.

### Redes Neuronales de Reservorio

Son redes neuronales recurrentes con una capa oculta fija y aleatoria, cuya salida se entrena linealmente. Son útiles en tareas de predicción temporal y reconocimiento de patrones.

### Redes Neuronales Autoorganizadas (SOM)

Las SOM son un tipo de red neuronal que utiliza el aprendizaje no supervisado para producir una representación de baja dimensión de los datos de entrada, a menudo en 2 dimensiones, que preserva la estructura topológica de los datos.

### MAS EJEMPLOS

- Redes Neuronales Convolucionales de Separación de Grupo (Grouped Convolutional Neural Networks): Estas son una variación de las CNNs donde los canales de entrada se dividen en grupos más pequeños, y cada grupo se convoluciona por separado. Esto puede reducir la cantidad de cálculos y permitir que las redes sean más profundas sin aumentar demasiado la complejidad computacional.

- Redes Neuronales de Memoria a Largo y Corto Plazo (Long Short-Term Memory, LSTM): Una subclase de las RNNs, las LSTM tienen "puertas" que permiten que la red recuerde y olvide información a lo largo del tiempo, lo que las hace muy eficaces para tratar con secuencias temporales de longitud variable.

- Redes Neuronales de Atención: Estas redes utilizan mecanismos de atención, a menudo en combinación con las RNNs o las CNNs, para permitir que la red se "centre" en diferentes partes de la entrada en diferentes momentos. Esto puede ser especialmente útil para tareas como la traducción automática o la generación de texto.

- Redes Neuronales de Hopfield: Son redes neuronales recurrentes completamente conectadas utilizadas para el reconocimiento de patrones. Pueden considerarse como uno de los primeros modelos de redes de memoria asociativa.

- Redes Neuronales de Base Radial (Radial Basis Function, RBF): Las RBF son un tipo de red neuronal de alimentación directa que utiliza funciones de base radial como funciones de activación. Son útiles para problemas de interpolación y aproximación de funciones.

- Redes Neuronales de Creencia Profunda (Deep Belief Networks, DBN): Las DBN son una clase de redes neuronales profundas que contienen múltiples capas de variables ocultas, con conexiones entre las capas pero no entre las unidades dentro de cada capa.

- Redes Neuronales de Máquinas de Vectores de Soporte (Support Vector Neural Networks): En este tipo de red, se aplica una función de kernel a la entrada antes de alimentarla a la red, lo que puede ayudar a manejar datos que no son linealmente separables.

- Redes Neuronales Difusas: Estas redes combinan la lógica difusa con las redes neuronales para manejar datos inciertos o vagos.

- Redes Neuronales de Transformadores: Introducidos en el paper "Attention is All You Need", los transformadores utilizan mecanismos de atención de auto-correlación para procesar secuencias de datos.

- Redes Neuronales de Variación Autoencoder (Variational Autoencoder, VAE): Las VAE son un tipo de red neuronal que aprende a comprimir los datos de entrada en un espacio latente codificado, y luego puede generar nuevas muestras de datos similares a los datos de entrada a partir de ese espacio latente.